Making EVE bootable by legacy BIOS

rvs opened the pull request:
This PR teaches EVE how to boot on a legacy BIOS. Beyond being useful for older PCs it is actually very handy with:
   * Coreboot boards
   * Google Compute Engine
   * Surprisingly enough it even simplifies Raspberry Pi bootloader process

The changes here seem to be pretty safe since on a UEFI based system (our primary focus in production) all of this gets simply ignored.

deitch commented on pkg/grub/Dockerfile:
```diff
@@ -19,32 +19,39 @@ RUN ln -s python3 /usr/bin/python
+  make install"
+
+ENV GRUB_PLATFORM=efi
+RUN eval $BUILD_GRUB
```
`eval`? Please don't.

rvs replied:
Well, the problem is -- I only want GRUB_PLATFORM to be expanded at the shell level, not at the Docker level. At the shell level I think eval is the only way to do this (unless you want me to do sh -c "$BUILD_GRUB" which is exactly what eval does -- but requires an extra copy of sh to be launched)

deitch commented on pkg/grub/Dockerfile:
```diff
@@ -19,32 +19,39 @@ RUN ln -s python3 /usr/bin/python
-  ./configure --libdir=/grub-lib --with-platform=efi CFLAGS="-Os -Wno-unused-value" && \
-  make -j "$(getconf _NPROCESSORS_ONLN)" && \
-  make install
+  ./configure --libdir=/grub-lib --with-platform=\${GRUB_PLATFORM} CFLAGS='-Os -Wno-unused-value' && \
```
`GRUB_PLATFORM` isn't set until 4 lines down.

rvs replied:
Yes -- that's on purpose -- see the comment above. I don't want this to be expanded at the Dockerfile variable expansion level -- I want it to be expanded at the shell level.

deitch commented on pkg/grub/Dockerfile:
```diff
@@ -19,32 +19,39 @@ RUN ln -s python3 /usr/bin/python
+
+ENV GRUB_PLATFORM=efi
+RUN eval $BUILD_GRUB
+RUN if [ "$(uname -m)" = x86_64 ]; then GRUB_PLATFORM="pc --disable-efiemu" && eval "$BUILD_GRUB" ; fi
```
This strikes me as a bit odd, but maybe I didn't understand. Are we not building a grub that can run on both x86_64 and i386 (and arm64)? Why are we checking the build machine's architecture?

rvs replied:
It is impossible to build GRUB that way -- you can build it either as a UEFI payload or as a legacy BIOS payload. Of course, legacy BIOS payload doesn't apply to ARM -- hence the check.

deitch commented:
As you said in the DM convo, @rvs , there is some "squirrelly logic" here. Two things would help a lot:

* a doc that described how grub is supposed to work to boot, or more correctly, the boot process across multiple architectures
* the logic is a little too convoluted for a `Dockerfile`. I think we need a separate command/script in the directory that gets copied in by the `Dockerfile` and then executed. That thing will be easier to understand and evaluate (and run on its own for testing).

deitch commented:
That doc should start by describing what we are trying to achieve: we want an EVE built image to be bootable in the following situations without modification:

* UEFI amd64
* UEFI arm64
* BIOS i386
* etc.

Here are the challenges, and here is how eve images are built to resolve them.

rvs commented:
Hey @deitch -- I think I addressed all your comments, refactored the shell implementation a little bit and provided quite a bit of documentation over here. https://github.com/rvs/eve/blob/bios/docs/BOOTING.md

I feel like this is ready to be merged now -- please take another look.
